File: sdk/dotnet/TerminalService.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;

namespace Rexec;

/// <summary>
/// Service for terminal WebSocket connections.
/// </summary>
public class TerminalService
{
    private readonly RexecClient _client;

    internal TerminalService(RexecClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Connect to a container's terminal.
    /// </summary>
    public Task<Terminal> ConnectAsync(string containerId, CancellationToken cancellationToken = default)
    {
        return ConnectAsync(containerId, 80, 24, cancellationToken);
    }

    /// <summary>
    /// Connect to a container's terminal with specified size.
    /// </summary>
    public async Task<Terminal> ConnectAsync(string containerId, int cols, int rows, CancellationToken cancellationToken = default)
    {
        var url = _client.GetWebSocketUrl($"/ws/terminal/{containerId}?cols={cols}&rows={rows}");
        var terminal = new Terminal(url, _client.Token);
        await terminal.ConnectAsync(cancellationToken);
        return terminal;
    }
}

/// <summary>
/// Interactive terminal connection to a container.
/// </summary>
public class Terminal : IAsyncDisposable
{
    private readonly ClientWebSocket _webSocket;
    private readonly string _url;
    private readonly string _token;
    private CancellationTokenSource? _receiveCts;
    private Task? _receiveTask;

    /// <summary>
    /// Event raised when data is received.
    /// </summary>
    public event Action<string>? OnData;

    /// <summary>
    /// Event raised when binary data is received.
    /// </summary>
    public event Action<byte[]>? OnBinaryData;

    /// <summary>
    /// Event raised when the connection closes.
    /// </summary>
    public event Action? OnClose;

    /// <summary>
    /// Event raised when an error occurs.
    /// </summary>
    public event Action<Exception>? OnError;

    /// <summary>
    /// Whether the terminal is connected.
    /// </summary>
    public bool IsConnected => _webSocket.State == WebSocketState.Open;

    internal Terminal(string url, string token)
    {
        _url = url;
        _token = token;
        _webSocket = new ClientWebSocket();
    }

    internal async Task ConnectAsync(CancellationToken cancellationToken)
    {
        _webSocket.Options.SetRequestHeader("Authorization", $"Bearer {_token}");

        try
        {
            await _webSocket.ConnectAsync(new Uri(_url), cancellationToken);
            _receiveCts = new CancellationTokenSource();
            _receiveTask = ReceiveLoopAsync(_receiveCts.Token);
        }
        catch (Exception ex)
        {
            throw new RexecException("Failed to connect to terminal", ex);
        }
    }

    /// <summary>
    /// Write data to the terminal.
    /// </summary>
    public async Task WriteAsync(string data, CancellationToken cancellationToken = default)
    {
        if (!IsConnected) return;

        var bytes = Encoding.UTF8.GetBytes(data);
        await _webSocket.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, true, cancellationToken);
    }

    /// <summary>
    /// Write binary data to the terminal.
    /// </summary>
    public async Task WriteAsync(byte[] data, CancellationToken cancellationToken = default)
    {
        if (!IsConnected) return;

        await _webSocket.SendAsync(new ArraySegment<byte>(data), WebSocketMessageType.Binary, true, cancellationToken);
    }

    /// <summary>
    /// Resize the terminal.
    /// </summary>
    public Task ResizeAsync(int cols, int rows, CancellationToken cancellationToken = default)
    {
        var resizeMsg = JsonSerializer.Serialize(new { type = "resize", cols, rows });
        return WriteAsync(resizeMsg, cancellationToken);
    }

    /// <summary>
    /// Close the terminal connection.
    /// </summary>
    public async Task CloseAsync(CancellationToken cancellationToken = default)
    {
        _receiveCts?.Cancel();

        if (_webSocket.State == WebSocketState.Open)
        {
            try
            {
                await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", cancellationToken);
            }
            catch { }
        }

        if (_receiveTask != null)
        {
            try
            {
                await _receiveTask;
            }
            catch { }
        }

        OnClose?.Invoke();
    }

    private async Task ReceiveLoopAsync(CancellationToken cancellationToken)
    {
        var buffer = new byte[4096];

        try
        {
            while (!cancellationToken.IsCancellationRequested && IsConnected)
            {
                var result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken);

                if (result.MessageType == WebSocketMessageType.Close)
                {
                    break;
                }

                var data = new byte[result.Count];
                Array.Copy(buffer, data, result.Count);

                if (result.MessageType == WebSocketMessageType.Text)
                {
                    OnData?.Invoke(Encoding.UTF8.GetString(data));
                }
                else
                {
                    OnBinaryData?.Invoke(data);
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (Exception ex)
        {
            OnError?.Invoke(ex);
        }
        finally
        {
            OnClose?.Invoke();
        }
    }

    public async ValueTask DisposeAsync()
    {
        await CloseAsync();
        _webSocket.Dispose();
        _receiveCts?.Dispose();
        GC.SuppressFinalize(this);
    }
}

File: sdk/php/src/TerminalService.php
<?php

declare(strict_types=1);

namespace Rexec;

use Ratchet\Client\WebSocket;
use Ratchet\Client\Connector;
use React\EventLoop\Loop;

/**
 * Service for terminal WebSocket connections.
 */
class TerminalService
{
    private RexecClient $client;

    public function __construct(RexecClient $client)
    {
        $this->client = $client;
    }

    /**
     * Connect to a container's terminal.
     *
     * @param string $containerId Container ID
     * @param int $cols Terminal columns
     * @param int $rows Terminal rows
     */
    public function connect(string $containerId, int $cols = 80, int $rows = 24): Terminal
    {
        $url = $this->client->getWebSocketUrl("/ws/terminal/{$containerId}?cols={$cols}&rows={$rows}");
        return new Terminal($url, $this->client->getToken());
    }
}

/**
 * Interactive terminal connection to a container.
 */
class Terminal
{
    private string $url;
    private string $token;
    private ?WebSocket $connection = null;
    private bool $connected = false;

    /** @var callable|null */
    private $onData = null;
    /** @var callable|null */
    private $onClose = null;
    /** @var callable|null */
    private $onError = null;

    public function __construct(string $url, string $token)
    {
        $this->url = $url;
        $this->token = $token;
    }

    /**
     * Open the WebSocket connection.
     *
     * @throws RexecException
     */
    public function open(): void
    {
        $loop = Loop::get();
        $connector = new Connector($loop);

        $connector($this->url, [], [
            'Authorization' => 'Bearer ' . $this->token,
        ])->then(
            function (WebSocket $conn) {
                $this->connection = $conn;
                $this->connected = true;

                $conn->on('message', function ($msg) {
                    if ($this->onData) {
                        ($this->onData)((string)$msg);
                    }
                });

                $conn->on('close', function () {
                    $this->connected = false;
                    if ($this->onClose) {
                        ($this->onClose)();
                    }
                });
            },
            function (\Exception $e) {
                if ($this->onError) {
                    ($this->onError)($e);
                }
                throw new RexecException('Failed to connect: ' . $e->getMessage());
            }
        );
    }

    /**
     * Write data to the terminal.
     */
    public function write(string $data): void
    {
        if ($this->connected && $this->connection) {
            $this->connection->send($data);
        }
    }

    /**
     * Resize the terminal.
     */
    public function resize(int $cols, int $rows): void
    {
        $this->write(json_encode([
            'type' => 'resize',
            'cols' => $cols,
            'rows' => $rows,
        ]));
    }

    /**
     * Set handler for data received.
     */
    public function onData(callable $handler): self
    {
        $this->onData = $handler;
        return $this;
    }

    /**
     * Set handler for close event.
     */
    public function onClose(callable $handler): self
    {
        $this->onClose = $handler;
        return $this;
    }

    /**
     * Set handler for errors.
     */
    public function onError(callable $handler): self
    {
        $this->onError = $handler;
        return $this;
    }

    /**
     * Check if the terminal is connected.
     */
    public function isConnected(): bool
    {
        return $this->connected;
    }

    /**
     * Close the terminal connection.
     */
    public function close(): void
    {
        $this->connected = false;
        if ($this->connection) {
            $this->connection->close();
            $this->connection = null;
        }
    }
}
